### PR TITLE
Expanding DataPipe  to support DataFrames

### PR DESCRIPTION
**Pull request by VitalyFedyunin**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #71931

Differential Revision: [D37500516](https://our.internmc.facebook.com/intern/diff/D37500516)

### CODE REVIEW

File: torch/utils/data/datapipes/iter/callable.py [hunk 1]
```diff
@@ -176,6 +208,18 @@ class CollatorIterDataPipe(MapperIterDataPipe):
     def __init__(
         self,
         datapipe: IterDataPipe,
-        collate_fn: Callable = _utils.collate.default_collate,
+        conversion: Optional[
+            Union[
+            Callable[[Any], Any],
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/callable.py:213]:
```suggestion
            Callable[..., Any],
```

File: torch/utils/data/datapipes/iter/callable.py [hunk 2]
```diff
@@ -131,6 +136,33 @@ def __setstate__(self, state):
             self.fn = dill_function  # type: ignore[assignment]
 
 
+def _collate_helper(conversion, item):
+    # TODO(VitalyFedyunin): Verify that item is any sort of batch
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/callable.py:140]: Do we need to check the type of `item` here?

File: torch/utils/data/datapipes/iter/callable.py [hunk 3]
```diff
@@ -131,6 +136,33 @@ def __setstate__(self, state):
...
+    if len(item.items) > 1:
+        print(item.items)
```
Comment (ejguan) [inline torch/utils/data/datapipes/iter/callable.py:142]: nit: removing print

And, I am curious where does the attribute `items` come from? What is the actual type of `item`?

File: torch/utils/data/datapipes/iter/callable.py (hunk 2 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/callable.py:140]: And, do we want to extend to support list of dictionary?

**Review by ejguan — APPROVED** [review]
LGTM with a few questions.

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37500516).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37500516).

File: torch/utils/data/datapipes/iter/callable.py (hunk 3 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/callable.py]: `DataChunkDF`

File: torch/utils/data/datapipes/iter/callable.py (hunk 2 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/callable.py:131]: As TODO line

File: torch/utils/data/datapipes/iter/callable.py (hunk 1 above)
Comment (VitalyFedyunin) [reply torch/utils/data/datapipes/iter/callable.py]: Done

**Comment by VitalyFedyunin** [comment]
Somehow internal tools lost approval. Need re-review.

**Review by ejguan — APPROVED** [review]

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37500516).

**Comment by VitalyFedyunin** [comment]
@VitalyFedyunin has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D37500516).

File: torch/utils/data/datapipes/dataframe/dataframes.py [hunk 4]
```diff
@@ -94,38 +124,127 @@ def __mul__(self, add_val):
...
+
+    def __exit__(self, *exc_info):
+        setattr(self.get_target(), self.attribute, self.save)
+
+
+class CaptureCall(Capture):
+
+    def __init__(self, callable, ctx=None, **kwargs):
+        if ctx is None:
+            self.ctx = {'operations': [], 'variables': []}
```
Comment (kit1980) [inline torch/utils/data/datapipes/dataframe/dataframes.py:209]: @VitalyFedyunin This seems to violate inheritance properties as base class `Capture` has more data. Usually you should call the parent's init before own init, but here it wouldn't really work. Many classes in this file have this issue.

**Pull request closed without merging**